### ISSUE TITLE
Node peers list requests, and handling of them

### ISSUE THREAD

**Issue by bladedoyle** [mimblewimble/grin#3561]
When my grin node starts it gets initial peers from dns and attempts to connect to each:
```
20210211 13:21:12.468 DEBUG grin_servers::grin::seed - Retrieving addresses from dns mainnet.seed.grin.icu:3414
20210211 13:21:12.516 DEBUG grin_servers::grin::seed - Retrieving addresses from dns mainnet.seed.713.mw:3414
...
20210211 13:21:13.080 TRACE grin_p2p::serv - connect_peer: on 0.0.0.0:3414. connecting to 157.230.142.11:3414
```
25 initial peer addresses are returned.

My node then adds the ones it can connect to as peers:
```
20210211 13:21:13.100 DEBUG grin_p2p::peer - connect: handshaking with Ok(157.230.142.11:3414)
20210211 13:21:13.117 DEBUG grin_p2p::store - save_peer: PeerAddr(157.230.142.11:3414) marked Healthy
```

All good so far.

For each of those peers my node then requests a peers list:
```
20210211 13:21:13.118 TRACE grin_p2p::peer - Asking 157.230.142.11:3414 for more peers PEER_LIST
```
Each peer returns a peer address list containing 256 entries.
My node then adds all of them to my peers database without checking to see if they are any good.
see: https://github.com/mimblewimble/grin/blob/298accb6f382ab9e96c73ea3fae70ab88d2372a9/p2p/src/peers.rs#L668

This could result in as many as 25 * 256 = 6400 peers added.
Each peer has to be added to the lmbd, then later almost all of them are found to be bad and removed.  Thats possibly many thousands of lmdb write lock requests within a second or two.
This flood of write lock requests overloads the lmdb and on a slow disk some peers db lock requests exceed the timeout and report "failed to get peers lock".

In practice some of the addresses in the peers lists are duplicates.
A quick test shows that 2741 (unique) peers are added to my peers database within the first second of starting a new node.
Of those 2741 peers added, only 57 accept a connection.
Within the first 30 seconds, 8357 (unique) peers are processed through my peers database.

**Comment by DavidBurkett** [comment]
Seems strange to add peers to the database if we've never verified their validity. Do we just do this because we don't have an in-memory mechanism for caching peers? Also, what's the reasoning for removing the addresses? Shouldn't we just mark the last time we tried to connect, and whether the connection succeeded? Otherwise we risk receiving the same addresses over and over, failing to connect to them, and deleting/forgetting we tried them. So when we receive a new peer list, we would end up needlessly trying those same peers again.

At the very least, we shouldn't be saving each bunch one-by-one. A simple improvement would be to change `add_peer` to `add_peers` and try saving all 256 as part of one db batch. I'm assuming that will just limit it to a single write lock.

What Grin++ does is keeps an in-memory cache of all the known peers. It reads the database on startup and then operates fully on the cache. Then there's a background task that syncs any updates to the database every 15-30 seconds. It does each update in a single batch, so it never has this locking issue.

**Comment by bladedoyle** [comment]
Instructions to reproduce - simulate a slow disk:

```
Create the disk file:
$ dd if=/dev/zero of=slow_disk bs=1G count=8

Create loopback device for it:
$ sodo losetup --show --find slow_disk
/dev/loop0

Get size:
$ sudo blockdev --getsize /dev/loop0
16777216

Map it to a disk device with N miliseconds "delay" both read and write (N=200):
$ echo "0 16777216 delay /dev/loop0 0 200" | sudo dmsetup create dm-slow_disk

Create a filesystem on it:
$ sudo mkfs -t ext4 /dev/mapper/dm-slow_disk

Mount it:
$ mkdir chain_data
$ sudo mount /dev/mapper/dm-slow_disk chain_data/
$ sudo chmod 777 chain_data/

Point the grin config to put chain_data there:
$ pwd
/home/bdoyle/dev/peers_lock
$ grin server config
grin-server.toml file configured and created in current directory
$ grep db_root grin-server.toml
db_root = "/home/bdoyle/dev/peers_lock/chain_data"
$ df -h /home/bdoyle/dev/peers_lock/chain_data
Filesystem                Size  Used Avail Use% Mounted on
/dev/mapper/dm-slow_disk  7.9G   36M  7.4G   1% /home/bdoyle/dev/peers_lock/chain_data


start grin:
$ grin

Observe peers locking errors on startup:
$ grep "peers lock" grin-server.log
20210215 11:28:20.165 ERROR grin_p2p::peers - connected_peers: failed to get peers lock
20210215 11:28:20.196 ERROR grin_p2p::peers - add_connected: failed to get peers lock
20210215 11:28:20.200 ERROR grin_p2p::peers - add_connected: failed to get peers lock
20210215 11:28:20.267 ERROR grin_p2p::peers - add_connected: failed to get peers lock
20210215 11:28:59.048 ERROR grin_p2p::peers - connected_peers: failed to get peers lock
...
...
```

**Comment by bladedoyle** [comment]
A second peers db locking bottleneck can be found in listen_for_addrs() in servers/src/grin/seed.rs

After adding thousands of peer addresses we got from requesting PEER_LISTs we spawn threads to test the connections and call peer.update_state() to record the result.  We are doing this in batches of "max_outbound_attempts = 128" every few seconds.  

On slow disk, this also causes peer lock contention and timeouts.

**Comment by antiochp** [comment]
Good catch @bladedoyle. Thanks for investigating this.

**Comment by antiochp** [comment]
Resolved in #3566.
